File: test/mydia/p2p/relay_list_test.exs
defmodule Mydia.P2p.RelayListTest do
  use ExUnit.Case, async: true

  alias Mydia.P2p.RelayList

  @default "https://cae1-1.relay.mydia.dev"
  @fetched "https://relay-one.example.test"
  @other "https://relay-two.example.test"

  setup do
    dir = Path.join(System.tmp_dir!(), "relay_list_test_#{System.unique_integer([:positive])}")
    File.mkdir_p!(dir)
    on_exit(fn -> File.rm_rf(dir) end)
    {:ok, cache_path: Path.join(dir, "relays.json")}
  end

  # Req accepts a plug as a 1-arity function, which is the cheapest way to
  # stand in for the relay without a socket.
  defp responds(status, body) do
    fn conn ->
      conn
      |> Plug.Conn.put_resp_content_type("application/json")
      |> Plug.Conn.send_resp(status, body)
    end
  end

  defp raises do
    fn _conn -> raise "connection refused" end
  end

  defp opts(cache_path, plug, extra \\ []) do
    Keyword.merge(
      [
        cache_path: cache_path,
        base_url: "https://relay.example.test",
        req_options: [plug: plug, retry: false],
        override: nil
      ],
      extra
    )
  end

  describe "override" do
    test "wins outright and makes no request", %{cache_path: cache_path} do
      plug = fn _conn -> raise "the override must not fetch" end

      assert {[@other], :override} =
               RelayList.resolve(opts(cache_path, plug, override: @other))
    end

    test "a blank override is treated as unset", %{cache_path: cache_path} do
      plug = responds(200, ~s({"p2p":{"relays":["#{@fetched}"]}}))

      assert {[@fetched], :fetched} =
               RelayList.resolve(opts(cache_path, plug, override: "   "))
    end

    test "a non-https override is rejected and falls through", %{cache_path: cache_path} do
      plug = responds(200, ~s({"p2p":{"relays":["#{@fetched}"]}}))

      assert {[@fetched], :fetched} =
               RelayList.resolve(opts(cache_path, plug, override: "ftp://nope.example.test"))
    end
  end

  describe "fetch" do
    test "uses the fetched list and writes the cache", %{cache_path: cache_path} do
      plug = responds(200, ~s({"p2p":{"relays":["#{@fetched}","#{@other}"]}}))

      assert {[@fetched, @other], :fetched} = RelayList.resolve(opts(cache_path, plug))
      assert Jason.decode!(File.read!(cache_path)) == [@fetched, @other]
    end

    test "drops entries that are not https URLs", %{cache_path: cache_path} do
      plug =
        responds(200, ~s({"p2p":{"relays":["http://insecure.test","not a url","#{@fetched}"]}}))

      assert {[@fetched], :fetched} = RelayList.resolve(opts(cache_path, plug))
    end

    test "ignores unknown keys in the document", %{cache_path: cache_path} do
      plug = responds(200, ~s({"future":{"thing":1},"p2p":{"relays":["#{@fetched}"],"extra":2}}))

      assert {[@fetched], :fetched} = RelayList.resolve(opts(cache_path, plug))
    end
  end

  describe "fallback" do
    test "a 500 falls back to the cache", %{cache_path: cache_path} do
      File.write!(cache_path, Jason.encode!([@other]))

      assert {[@other], :cached} = RelayList.resolve(opts(cache_path, responds(500, "nope")))
    end

    test "a transport failure falls back to the cache", %{cache_path: cache_path} do
      File.write!(cache_path, Jason.encode!([@other]))

      assert {[@other], :cached} = RelayList.resolve(opts(cache_path, raises()))
    end

    test "malformed JSON falls back to the cache", %{cache_path: cache_path} do
      File.write!(cache_path, Jason.encode!([@other]))

      assert {[@other], :cached} = RelayList.resolve(opts(cache_path, responds(200, "{{{")))
    end

    test "an empty list after filtering falls back to the cache", %{cache_path: cache_path} do
      File.write!(cache_path, Jason.encode!([@other]))
      plug = responds(200, ~s({"p2p":{"relays":["http://insecure.test"]}}))

      assert {[@other], :cached} = RelayList.resolve(opts(cache_path, plug))
    end

    test "no cache falls back to the compiled-in default", %{cache_path: cache_path} do
      assert {[@default], :default} = RelayList.resolve(opts(cache_path, responds(500, "nope")))
    end

    test "an unreadable cache falls back to the compiled-in default", %{cache_path: cache_path} do
      File.write!(cache_path, "{{{ not json")

      assert {[@default], :default} = RelayList.resolve(opts(cache_path, responds(500, "nope")))
    end

    test "a failed fetch does not overwrite a good cache", %{cache_path: cache_path} do
      File.write!(cache_path, Jason.encode!([@other]))

      RelayList.resolve(opts(cache_path, responds(500, "nope")))

      assert Jason.decode!(File.read!(cache_path)) == [@other]
    end
  end

  describe "default_relay_url/0" do
    test "is the relay compiled into this build" do
      assert RelayList.default_relay_url() == @default
    end
  end
end

File: player/test/core/p2p/relay_list_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:player/core/auth/auth_storage.dart';
import 'package:player/core/p2p/relay_list.dart';

/// An in-memory AuthStorage so the resolver's cache can be asserted without a
/// platform channel.
class _MemoryStorage implements AuthStorage {
  final Map<String, String> values = {};

  @override
  bool get degraded => false;

  @override
  Future<String?> read(String key) async => values[key];

  @override
  Future<void> write(String key, String value) async => values[key] = value;

  @override
  Future<void> delete(String key) async => values.remove(key);

  @override
  Future<void> deleteAll() async => values.clear();
}

const _fetched = 'https://relay-one.example.test';
const _other = 'https://relay-two.example.test';

http.Client _responds(int status, String body) =>
    MockClient((_) async => http.Response(body, status));

http.Client _fails() =>
    MockClient((_) async => throw http.ClientException('refused'));

http.Client _neverCalled() =>
    MockClient((_) async => fail('no request should have been made'));

void main() {
  group('override', () {
    test('wins outright and makes no request', () async {
      final result = await resolveRelayList(
        override: _other,
        client: _neverCalled(),
        storage: _MemoryStorage(),
      );

      expect(result.urls, [_other]);
      expect(result.source, RelayListSource.override);
    });

    test('a blank override is treated as unset', () async {
      final result = await resolveRelayList(
        override: '   ',
        client: _responds(
            200,
            jsonEncode({
              'p2p': {
                'relays': [_fetched]
              }
            })),
        storage: _MemoryStorage(),
      );

      expect(result.urls, [_fetched]);
      expect(result.source, RelayListSource.fetched);
    });

    test('a non-https override falls through', () async {
      final result = await resolveRelayList(
        override: 'ftp://nope.example.test',
        client: _responds(
            200,
            jsonEncode({
              'p2p': {
                'relays': [_fetched]
              }
            })),
        storage: _MemoryStorage(),
      );

      expect(result.source, RelayListSource.fetched);
    });
  });

  group('fetch', () {
    test('uses the fetched list and caches it', () async {
      final storage = _MemoryStorage();

      final result = await resolveRelayList(
        client: _responds(
            200,
            jsonEncode({
              'p2p': {
                'relays': [_fetched, _other]
              }
            })),
        storage: storage,
      );

      expect(result.urls, [_fetched, _other]);
      expect(result.source, RelayListSource.fetched);
      expect(
          jsonDecode(storage.values[relayListStorageKey]!), [_fetched, _other]);
    });

    test('drops entries that are not https URLs', () async {
      final result = await resolveRelayList(
        client: _responds(
            200,
            jsonEncode({
              'p2p': {
                'relays': ['http://insecure.test', 'not a url', _fetched]
              }
            })),
        storage: _MemoryStorage(),
      );

      expect(result.urls, [_fetched]);
    });

    test('ignores unknown keys', () async {
      final result = await resolveRelayList(
        client: _responds(
            200,
            jsonEncode({
              'future': {'thing': 1},
              'p2p': {
                'relays': [_fetched],
                'extra': 2
              }
            })),
        storage: _MemoryStorage(),
      );

      expect(result.urls, [_fetched]);
      expect(result.source, RelayListSource.fetched);
    });
  });

  group('fallback', () {
    Future<RelayListResult> withCache(http.Client client) {
      final storage = _MemoryStorage();
      storage.values[relayListStorageKey] = jsonEncode([_other]);
      return resolveRelayList(client: client, storage: storage);
    }

    test('a 500 falls back to the cache', () async {
      final result = await withCache(_responds(500, 'nope'));
      expect(result.urls, [_other]);
      expect(result.source, RelayListSource.cached);
    });

    test('a transport failure falls back to the cache', () async {
      final result = await withCache(_fails());
      expect(result.urls, [_other]);
      expect(result.source, RelayListSource.cached);
    });

    test('malformed JSON falls back to the cache', () async {
      final result = await withCache(_responds(200, '{{{'));
      expect(result.urls, [_other]);
      expect(result.source, RelayListSource.cached);
    });

    test('an empty list after filtering falls back to the cache', () async {
      final result = await withCache(_responds(
          200,
          jsonEncode({
            'p2p': {
              'relays': ['http://insecure.test']
            }
          })));
      expect(result.urls, [_other]);
      expect(result.source, RelayListSource.cached);
    });

    test('no cache falls back to the compiled-in default', () async {
      final result = await resolveRelayList(
        client: _responds(500, 'nope'),
        storage: _MemoryStorage(),
      );

      expect(result.urls, [defaultIrohRelayUrl]);
      expect(result.source, RelayListSource.defaultBuiltIn);
    });

    test('an unreadable cache falls back to the compiled-in default', () async {
      final storage = _MemoryStorage();
      storage.values[relayListStorageKey] = '{{{ not json';

      final result = await resolveRelayList(
        client: _responds(500, 'nope'),
        storage: storage,
      );

      expect(result.urls, [defaultIrohRelayUrl]);
      expect(result.source, RelayListSource.defaultBuiltIn);
    });

    test('a failed fetch does not overwrite a good cache', () async {
      final storage = _MemoryStorage();
      storage.values[relayListStorageKey] = jsonEncode([_other]);

      await resolveRelayList(client: _responds(500, 'nope'), storage: storage);

      expect(jsonDecode(storage.values[relayListStorageKey]!), [_other]);
    });
  });
}
